bytecomp/matching.ml

```diff
@@ -2836,51 +2874,47 @@ let find_in_pat pred =
   end in
   find_rec
 
-let is_lazy_pat = function
-  | Tpat_lazy _ -> true
-  | Tpat_alias _ | Tpat_variant _ | Tpat_record _
-  | Tpat_tuple _|Tpat_construct _ | Tpat_array _
-  | Tpat_or _ | Tpat_constant _ | Tpat_var _ | Tpat_any
-      -> false
-
-let is_lazy p = find_in_pat is_lazy_pat p
-
-let have_mutable_field p = match p with
+let is_mutable_pat p = match p with
 | Tpat_record (lps,_) ->
     List.exists
       (fun (_,lbl,_) ->
         match lbl.Types.lbl_mut with
         | Mutable -> true
         | Immutable -> false)
       lps
+| Tpat_array _ -> true
 | Tpat_alias _ | Tpat_variant _ | Tpat_lazy _
-| Tpat_tuple _|Tpat_construct _ | Tpat_array _
+| Tpat_tuple _|Tpat_construct _
 | Tpat_or _
 | Tpat_constant _ | Tpat_var _ | Tpat_any
   -> false
 
-let is_mutable p = find_in_pat have_mutable_field p
+let is_mutable p = find_in_pat is_mutable_pat p
 
 (* Downgrade Total when
    1. Matching accesses some mutable fields;
    2. And there are  guards or lazy patterns.
```

Comment:
The new downgrading implementation does not take guards/lazy into account anymore, does it?